File: src/test/java/no/fintlabs/flyt/gateway/application/archive/dispatch/DispatchServiceTest.java
package no.fintlabs.flyt.gateway.application.archive.dispatch;

import no.fintlabs.flyt.gateway.application.archive.dispatch.model.CaseDispatchType;
import no.fintlabs.flyt.gateway.application.archive.dispatch.model.instance.ArchiveInstance;
import no.fintlabs.flyt.gateway.application.archive.dispatch.model.instance.JournalpostDto;
import no.fintlabs.flyt.gateway.application.archive.dispatch.model.instance.SakDto;
import no.fintlabs.flyt.gateway.application.archive.dispatch.sak.CaseDispatchService;
import no.fintlabs.flyt.gateway.application.archive.dispatch.sak.result.CaseDispatchResult;
import no.fintlabs.flyt.gateway.application.archive.dispatch.sak.result.CaseSearchResult;
import no.fintlabs.flyt.kafka.headers.InstanceFlowHeaders;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

import java.util.List;
import java.util.Optional;

import static no.fintlabs.flyt.gateway.application.archive.dispatch.DispatchStatus.ACCEPTED;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class DispatchServiceTest {

    @Mock
    private CaseDispatchService caseDispatchService;
    @Mock
    private InstanceFlowHeaders instanceFlowHeaders;
    @Mock
    private ArchiveInstance archiveInstance;
    @Mock
    private RecordsProcessingService recordsProcessingService;
    @InjectMocks
    private DispatchService dispatchService;

    @Test
    public void givenCaseTypeNewAndAcceptedCaseDispatchWithNoJournalpostShouldReturnAcceptedResultWithCaseId() {
        doReturn(CaseDispatchType.NEW).when(archiveInstance).getType();

        SakDto sakDto = mock(SakDto.class);
        doReturn(sakDto).when(archiveInstance).getNewCase();
        doReturn(Mono.just(CaseDispatchResult.accepted("testCaseId")))
                .when(caseDispatchService).dispatch(sakDto);

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.accepted("testCaseId"))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(2)).getNewCase();
        verifyNoMoreInteractions(archiveInstance);

        verify(caseDispatchService, times(1)).dispatch(sakDto);
        verifyNoMoreInteractions(caseDispatchService);
    }

    @Test
    public void givenCaseTypeNewAndAcceptedCaseDispatchShouldCallRecordsProcessingServiceAndReturnResult() {
        doReturn(CaseDispatchType.NEW).when(archiveInstance).getType();

        SakDto sakDto = mock(SakDto.class);
        doReturn(sakDto).when(archiveInstance).getNewCase();
        doReturn(Mono.just(CaseDispatchResult.accepted("testCaseId")))
                .when(caseDispatchService).dispatch(sakDto);

        JournalpostDto journalpostDto = mock(JournalpostDto.class);
        doReturn(Optional.of(List.of(journalpostDto))).when(sakDto).getJournalpost();

        doReturn(Mono.just(DispatchResult.accepted("testCaseId")))
                .when(recordsProcessingService)
                .processRecords("testCaseId", true, List.of(journalpostDto));

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.accepted("testCaseId"))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(2)).getNewCase();
        verifyNoMoreInteractions(archiveInstance);

        verify(caseDispatchService, times(1)).dispatch(sakDto);
        verifyNoMoreInteractions(caseDispatchService);

        verify(recordsProcessingService, times(1)).
                processRecords("testCaseId", true, List.of(journalpostDto));
        verifyNoMoreInteractions(recordsProcessingService);
    }

    @Test
    public void givenCaseTypeNewAndDeclinedCaseDispatchShouldReturnDeclinedResultWithErrorMessage() {
        doReturn(CaseDispatchType.NEW).when(archiveInstance).getType();

        SakDto sakDto = mock(SakDto.class);
        doReturn(sakDto).when(archiveInstance).getNewCase();

        doReturn(Mono.just(CaseDispatchResult.declined("test error message")))
                .when(caseDispatchService)
                .dispatch(sakDto);

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult
                        .declined("Sak was declined by the destination with message='test error message'")
                )
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(1)).getNewCase();
        verifyNoMoreInteractions(archiveInstance);

        verify(caseDispatchService, times(1)).dispatch(sakDto);
        verifyNoMoreInteractions(caseDispatchService);
    }

    @Test
    public void givenCaseTypeNewAndFailedCaseDispatchShouldReturnFailedResultWithErrorMessage() {
        doReturn(CaseDispatchType.NEW).when(archiveInstance).getType();

        SakDto sakDto = mock(SakDto.class);
        doReturn(sakDto).when(archiveInstance).getNewCase();

        doReturn(Mono.just(CaseDispatchResult.failed())).when(caseDispatchService).dispatch(sakDto);

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.failed("Sak dispatch failed"))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(1)).getNewCase();
        verifyNoMoreInteractions(archiveInstance);

        verify(caseDispatchService, times(1)).dispatch(sakDto);
        verifyNoMoreInteractions(caseDispatchService);
    }

    @Test
    public void givenCaseTypeByIdAndAcceptedJournalpostDispatchShouldCallRecordsProcessingServiceWithNewCaseFalseAndReturnResult() {
        doReturn(CaseDispatchType.BY_ID).when(archiveInstance).getType();
        doReturn("testCaseId").when(archiveInstance).getCaseId();

        JournalpostDto journalpostDto = mock(JournalpostDto.class);
        doReturn(List.of(journalpostDto)).when(archiveInstance).getJournalpost();

        doReturn(Mono.just(DispatchResult.accepted("testCaseId")))
                .when(recordsProcessingService)
                .processRecords("testCaseId", false, List.of(journalpostDto));

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.accepted("testCaseId"))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(1)).getCaseId();
        verifyNoMoreInteractions(archiveInstance);

        verify(recordsProcessingService, times(1))
                .processRecords("testCaseId", false, List.of(journalpostDto));
        verifyNoMoreInteractions(recordsProcessingService);
    }

    @Test
    public void givenCaseTypeByIdAndDeclinedJournalpostDispatchShouldCallRecordsProcessingServiceWithNewCaseFalseAndReturnResult() {
        doReturn(CaseDispatchType.BY_ID).when(archiveInstance).getType();
        doReturn("testCaseId").when(archiveInstance).getCaseId();

        JournalpostDto journalpostDto = mock(JournalpostDto.class);
        doReturn(List.of(journalpostDto)).when(archiveInstance).getJournalpost();

        doReturn(Mono.just(DispatchResult.declined("test error message")))
                .when(recordsProcessingService)
                .processRecords("testCaseId", false, List.of(journalpostDto));

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.declined(
                        "test error message"
                ))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(1)).getCaseId();
        verifyNoMoreInteractions(archiveInstance);

        verify(recordsProcessingService, times(1))
                .processRecords("testCaseId", false, List.of(journalpostDto));
        verifyNoMoreInteractions(recordsProcessingService);
    }

    @Test
    public void givenCaseByIdAndFailedJournalpostDispatchShouldCallRecordsProcessingServiceWithNewCaseFalseAndReturnResult() {
        doReturn(CaseDispatchType.BY_ID).when(archiveInstance).getType();
        doReturn("testCaseId").when(archiveInstance).getCaseId();

        JournalpostDto journalpostDto = mock(JournalpostDto.class);
        doReturn(List.of(journalpostDto)).when(archiveInstance).getJournalpost();

        doReturn(Mono.just(DispatchResult.failed("test error message")))
                .when(recordsProcessingService)
                .processRecords("testCaseId", false, List.of(journalpostDto));

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.failed("test error message"
                ))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(1)).getCaseId();
        verifyNoMoreInteractions(archiveInstance);

        verify(recordsProcessingService, times(1))
                .processRecords("testCaseId", false, List.of(journalpostDto));
        verifyNoMoreInteractions(recordsProcessingService);
    }

    @Test
    public void givenCaseTypeBySearchOrNewWithJournalPostShouldReturnDeclinedResultWhenMultipleCasesAreFound() {
        doReturn(CaseDispatchType.BY_SEARCH_OR_NEW).when(archiveInstance).getType();

        SakDto sakDto = mock(SakDto.class);
        doReturn(sakDto).when(archiveInstance).getNewCase();

        JournalpostDto journalpostDto = mock(JournalpostDto.class);
        doReturn(Optional.of(List.of(journalpostDto))).when(sakDto).getJournalpost();

        CaseSearchResult caseSearchResult = mock(CaseSearchResult.class);
        doReturn(List.of("caseId1", "caseId2")).when(caseSearchResult).getArchiveCaseIds();
        doReturn(ACCEPTED).when(caseSearchResult).getStatus();

        doReturn(Mono.just(caseSearchResult)).when(caseDispatchService).findCasesBySearch(archiveInstance);

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.declined("Found multiple cases: caseId1, caseId2"))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(1)).getNewCase();
        verifyNoMoreInteractions(archiveInstance);

        verify(caseDispatchService, times(1)).findCasesBySearch(archiveInstance);
        verifyNoMoreInteractions(caseDispatchService);
    }

    @Test
    public void givenCaseTypeBySearchOrNewWithJournalPostAndOneCaseIsFoundAndAcceptedJournalpostDispatchShouldCallRecordsProcessingServiceWithNewCaseFalseAndReturnResult() {
        doReturn(CaseDispatchType.BY_SEARCH_OR_NEW).when(archiveInstance).getType();

        SakDto sakDto = mock(SakDto.class);
        doReturn(sakDto).when(archiveInstance).getNewCase();

        JournalpostDto journalpostDto = mock(JournalpostDto.class);
        doReturn(Optional.of(List.of(journalpostDto))).when(sakDto).getJournalpost();

        CaseSearchResult caseSearchResult = mock(CaseSearchResult.class);
        doReturn(List.of("testCaseId")).when(caseSearchResult).getArchiveCaseIds();
        doReturn(ACCEPTED).when(caseSearchResult).getStatus();

        doReturn(Mono.just(caseSearchResult)).when(caseDispatchService).findCasesBySearch(archiveInstance);

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.accepted("testCaseId"))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(1)).getNewCase();
        verifyNoMoreInteractions(archiveInstance);

        verify(caseDispatchService, times(1)).findCasesBySearch(archiveInstance);
        verifyNoMoreInteractions(caseDispatchService);

        verify(recordsProcessingService, times(1))
                .processRecords("testCaseId", false, List.of(journalpostDto));
        verifyNoMoreInteractions(recordsProcessingService);
    }

    @Test
    public void givenCaseTypeBySearchOrNewWithJournalPostAndNoCaseFoundShouldCallRecordsProcessingServiceWithNewCaseTrueAndReturnResult() {
        doReturn(CaseDispatchType.BY_SEARCH_OR_NEW).when(archiveInstance).getType();

        SakDto sakDto = mock(SakDto.class);
        doReturn(sakDto).when(archiveInstance).getNewCase();

        JournalpostDto journalpostDto = mock(JournalpostDto.class);
        doReturn(Optional.of(List.of(journalpostDto))).when(sakDto).getJournalpost();

        doReturn(Mono.just(CaseDispatchResult.accepted("testCaseId"))).when(caseDispatchService)
                .dispatch(sakDto);

        CaseSearchResult caseSearchResult = mock(CaseSearchResult.class);
        doReturn(List.of()).when(caseSearchResult).getArchiveCaseIds();
        doReturn(ACCEPTED).when(caseSearchResult).getStatus();

        doReturn(Mono.just(caseSearchResult)).when(caseDispatchService).findCasesBySearch(archiveInstance);

        doReturn(Mono.just(DispatchResult.accepted("testCaseId")))
                .when(recordsProcessingService)
                .processRecords("testCaseId", true, List.of(journalpostDto));

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.accepted("testCaseId"))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(3)).getNewCase();
        verifyNoMoreInteractions(archiveInstance);

        verify(caseDispatchService, times(1)).findCasesBySearch(archiveInstance);
        verifyNoMoreInteractions(caseDispatchService);

        verify(recordsProcessingService, times(1))
                .processRecords("testCaseId", true, List.of(journalpostDto));
        verifyNoMoreInteractions(recordsProcessingService);
    }

    @Test
    public void givenCaseTypeBySearchOrNewAndDeclinedCaseSearchShouldReturnDeclinedResult() {
        doReturn(CaseDispatchType.BY_SEARCH_OR_NEW).when(archiveInstance).getType();

        SakDto sakDto = mock(SakDto.class);
        doReturn(sakDto).when(archiveInstance).getNewCase();
        doReturn(Optional.of(List.of())).when(sakDto).getJournalpost();

        CaseSearchResult caseSearchResult = mock(CaseSearchResult.class);
        doReturn(DispatchStatus.DECLINED).when(caseSearchResult).getStatus();
        doReturn("test errormessage").when(caseSearchResult).getErrorMessage();

        doReturn(Mono.just(caseSearchResult)).when(caseDispatchService).findCasesBySearch(archiveInstance);

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.declined("test errormessage"))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(1)).getNewCase();
        verifyNoMoreInteractions(archiveInstance);

        verify(caseDispatchService, times(1)).findCasesBySearch(archiveInstance);
        verifyNoMoreInteractions(caseDispatchService);
    }

    @Test
    public void givenCaseTypeBySearchOrNewAndFailedCaseSearchShouldReturnFailedResult() {
        doReturn(CaseDispatchType.BY_SEARCH_OR_NEW).when(archiveInstance).getType();

        SakDto sakDto = mock(SakDto.class);
        doReturn(sakDto).when(archiveInstance).getNewCase();
        doReturn(Optional.of(List.of())).when(sakDto).getJournalpost();

        CaseSearchResult caseSearchResult = mock(CaseSearchResult.class);
        doReturn(DispatchStatus.FAILED).when(caseSearchResult).getStatus();

        doReturn(Mono.just(caseSearchResult)).when(caseDispatchService).findCasesBySearch(archiveInstance);

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.failed())
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(1)).getNewCase();
        verifyNoMoreInteractions(archiveInstance);

        verify(caseDispatchService, times(1)).findCasesBySearch(archiveInstance);
        verifyNoMoreInteractions(caseDispatchService);
    }

    @Test
    public void givenCaseTypeBySearchOrNewAndNoCaseFoundAndNewCaseDeclinedShouldReturnDeclinedResult() {
        doReturn(CaseDispatchType.BY_SEARCH_OR_NEW).when(archiveInstance).getType();

        SakDto sakDto = mock(SakDto.class);
        doReturn(sakDto).when(archiveInstance).getNewCase();
        doReturn(Optional.of(List.of())).when(sakDto).getJournalpost();

        CaseSearchResult caseSearchResult = mock(CaseSearchResult.class);
        doReturn(ACCEPTED).when(caseSearchResult).getStatus();
        doReturn(List.of()).when(caseSearchResult).getArchiveCaseIds();
        doReturn(Mono.just(caseSearchResult)).when(caseDispatchService).findCasesBySearch(archiveInstance);

        CaseDispatchResult caseDispatchResult = mock(CaseDispatchResult.class);
        doReturn(DispatchStatus.DECLINED).when(caseDispatchResult).getStatus();
        doReturn("testErrorMessage").when(caseDispatchResult).getErrorMessage();
        doReturn(Mono.just(caseDispatchResult)).when(caseDispatchService).dispatch(sakDto);

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.declined("Sak was declined by the destination with message='testErrorMessage'"))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(2)).getNewCase();
        verifyNoMoreInteractions(archiveInstance);

        verify(caseDispatchService, times(1)).findCasesBySearch(archiveInstance);
        verify(caseDispatchService, times(1)).dispatch(sakDto);
        verifyNoMoreInteractions(caseDispatchService);
    }

    @Test
    public void givenCaseTypeBySearchOrNewAndNoCaseFoundAndNewCaseFailedShouldReturnFailedResult() {
        doReturn(CaseDispatchType.BY_SEARCH_OR_NEW).when(archiveInstance).getType();

        SakDto sakDto = mock(SakDto.class);
        doReturn(sakDto).when(archiveInstance).getNewCase();
        doReturn(Optional.of(List.of())).when(sakDto).getJournalpost();

        CaseSearchResult caseSearchResult = mock(CaseSearchResult.class);
        doReturn(ACCEPTED).when(caseSearchResult).getStatus();
        doReturn(List.of()).when(caseSearchResult).getArchiveCaseIds();
        doReturn(Mono.just(caseSearchResult)).when(caseDispatchService).findCasesBySearch(archiveInstance);

        CaseDispatchResult caseDispatchResult = mock(CaseDispatchResult.class);
        doReturn(DispatchStatus.FAILED).when(caseDispatchResult).getStatus();
        doReturn(Mono.just(caseDispatchResult)).when(caseDispatchService).dispatch(sakDto);

        StepVerifier.create(
                        dispatchService.process(instanceFlowHeaders, archiveInstance)
                )
                .expectNext(DispatchResult.failed("Sak dispatch failed"))
                .verifyComplete();

        verify(archiveInstance, times(1)).getType();
        verify(archiveInstance, times(2)).getNewCase();
        verifyNoMoreInteractions(archiveInstance);

        verify(caseDispatchService, times(1)).findCasesBySearch(archiveInstance);
        verify(caseDispatchService, times(1)).dispatch(sakDto);
        verifyNoMoreInteractions(caseDispatchService);
    }

}

File: src/test/kotlin/no/novari/flyt/archive/gateway/dispatch/DispatchServiceTest.kt
package no.novari.flyt.archive.gateway.dispatch

import no.novari.flyt.archive.gateway.dispatch.mapping.DokumentetsDatoMappingService
import no.novari.flyt.archive.gateway.dispatch.model.CaseDispatchType
import no.novari.flyt.archive.gateway.dispatch.model.instance.ArchiveInstance
import no.novari.flyt.archive.gateway.dispatch.model.instance.JournalpostDto
import no.novari.flyt.archive.gateway.dispatch.model.instance.SakDto
import no.novari.flyt.archive.gateway.dispatch.sak.CaseDispatchService
import no.novari.flyt.archive.gateway.dispatch.sak.result.CaseDispatchResult
import no.novari.flyt.archive.gateway.dispatch.sak.result.CaseSearchResult
import no.novari.flyt.kafka.instanceflow.headers.InstanceFlowHeaders
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoInteractions
import org.mockito.kotlin.whenever

@ExtendWith(MockitoExtension::class)
class DispatchServiceTest {
    @Mock
    private lateinit var caseDispatchService: CaseDispatchService

    @Mock
    private lateinit var instanceFlowHeaders: InstanceFlowHeaders

    @Mock
    private lateinit var recordsProcessingService: RecordsProcessingService

    private val dokumentetsDatoMappingService = DokumentetsDatoMappingService()

    private lateinit var dispatchService: DispatchService

    @BeforeEach
    fun setup() {
        dispatchService =
            DispatchService(
                caseDispatchService,
                recordsProcessingService,
                dokumentetsDatoMappingService,
            )
    }

    @Test
    fun `given case type NEW with no journalpost and an accepted dispatch, returns an accepted result with case id`() {
        val sakDto = SakDto.builder().build()
        val archiveInstance =
            ArchiveInstance
                .builder()
                .type(CaseDispatchType.NEW)
                .newCase(sakDto)
                .build()
        whenever(caseDispatchService.dispatch(sakDto)).thenReturn(CaseDispatchResult.accepted("testCaseId"))

        val result = dispatchService.process(instanceFlowHeaders, archiveInstance)

        assertThat(result).isEqualTo(DispatchResult.accepted("testCaseId"))
    }

    @Test
    fun `given case type NEW and an accepted dispatch, calls records processing service and returns its result`() {
        val journalpostDto = mock<JournalpostDto>()
        val sakDto = SakDto.builder().journalpost(listOf(journalpostDto)).build()
        val archiveInstance =
            ArchiveInstance
                .builder()
                .type(CaseDispatchType.NEW)
                .newCase(sakDto)
                .build()
        whenever(caseDispatchService.dispatch(sakDto)).thenReturn(CaseDispatchResult.accepted("testCaseId"))
        whenever(recordsProcessingService.processRecords("testCaseId", true, listOf(journalpostDto)))
            .thenReturn(DispatchResult.accepted("testCaseId"))

        val result = dispatchService.process(instanceFlowHeaders, archiveInstance)

        assertThat(result).isEqualTo(DispatchResult.accepted("testCaseId"))
    }

    @Test
    fun `given case type BY_ID, calls records processing service with newCase=false`() {
        val journalpostDto = mock<JournalpostDto>()
        val archiveInstance =
            ArchiveInstance
                .builder()
                .type(CaseDispatchType.BY_ID)
                .caseId("testCaseId")
                .journalpost(listOf(journalpostDto))
                .build()
        whenever(recordsProcessingService.processRecords("testCaseId", false, listOf(journalpostDto)))
            .thenReturn(DispatchResult.accepted("testCaseId"))

        val result = dispatchService.process(instanceFlowHeaders, archiveInstance)

        assertThat(result).isEqualTo(DispatchResult.accepted("testCaseId"))
    }

    @Test
    fun `given case type BY_SEARCH_OR_NEW with multiple matching cases, returns a declined result`() {
        val journalpostDto = mock<JournalpostDto>()
        val sakDto = SakDto.builder().journalpost(listOf(journalpostDto)).build()
        val archiveInstance =
            ArchiveInstance
                .builder()
                .type(CaseDispatchType.BY_SEARCH_OR_NEW)
                .newCase(sakDto)
                .build()
        whenever(caseDispatchService.findCasesBySearch(archiveInstance))
            .thenReturn(CaseSearchResult.accepted(listOf("caseId1", "caseId2")))

        val result = dispatchService.process(instanceFlowHeaders, archiveInstance)

        assertThat(result).isEqualTo(DispatchResult.declined("Found multiple cases: caseId1, caseId2"))
    }

    @Test
    fun `given case type BY_SEARCH_OR_NEW and no cases found, delegates to processNew`() {
        val sakDto = SakDto.builder().build()
        val archiveInstance =
            ArchiveInstance
                .builder()
                .type(CaseDispatchType.BY_SEARCH_OR_NEW)
                .newCase(sakDto)
                .build()
        whenever(caseDispatchService.findCasesBySearch(archiveInstance))
            .thenReturn(CaseSearchResult.accepted(emptyList()))
        whenever(caseDispatchService.dispatch(sakDto)).thenReturn(CaseDispatchResult.accepted("testCaseId"))

        val result = dispatchService.process(instanceFlowHeaders, archiveInstance)

        assertThat(result).isEqualTo(DispatchResult.accepted("testCaseId"))
    }

    @Test
    fun `given invalid dokumentetsDato for new case, returns declined result before dispatching case`() {
        val sakDto =
            SakDto
                .builder()
                .journalpost(listOf(JournalpostDto.builder().dokumentetsDato("1990-02-30").build()))
                .build()
        val archiveInstance =
            ArchiveInstance
                .builder()
                .type(CaseDispatchType.NEW)
                .newCase(sakDto)
                .build()

        val result = dispatchService.process(instanceFlowHeaders, archiveInstance)

        assertThat(result).isEqualTo(
            DispatchResult.declined(
                "Invalid dokumentetsDato='1990-02-30'. Expected ISO 8601 date " +
                    "'yyyy-MM-dd' or date-time 'yyyy-MM-dd'T'HH:mm:ssZ'.",
            ),
        )
        verify(caseDispatchService, never()).dispatch(sakDto)
        verifyNoInteractions(recordsProcessingService)
    }

    @Test
    fun `given invalid dokumentetsDato for existing case, returns declined result before dispatching journalpost`() {
        val journalpostDto = JournalpostDto.builder().dokumentetsDato("not-a-date").build()
        val archiveInstance =
            ArchiveInstance
                .builder()
                .type(CaseDispatchType.BY_ID)
                .caseId("testCaseId")
                .journalpost(listOf(journalpostDto))
                .build()

        val result = dispatchService.process(instanceFlowHeaders, archiveInstance)

        assertThat(result).isEqualTo(
            DispatchResult.declined(
                "Invalid dokumentetsDato='not-a-date'. Expected ISO 8601 date " +
                    "'yyyy-MM-dd' or date-time 'yyyy-MM-dd'T'HH:mm:ssZ'.",
            ),
        )
        verifyNoInteractions(caseDispatchService)
        verify(recordsProcessingService, never()).processRecords("testCaseId", false, listOf(journalpostDto))
    }
}
